File: src/uls/ulst.f90
! THIS VERSION: GALAHAD 2.4 - 19/06/2009 AT 12:00 GMT.
   PROGRAM GALAHAD_ULS_TEST_PROGRAM
   USE GALAHAD_SYMBOLS
   USE GALAHAD_ULS_double
   IMPLICIT NONE
   INTEGER, PARAMETER :: wp = KIND( 1.0D+0 )
   TYPE ( SMT_type ) :: matrix
   TYPE ( ULS_data_type ) :: data
   TYPE ( ULS_control_type ) control
   TYPE ( ULS_inform_type ) :: inform
   INTEGER :: i, ordering, solver, type, s
   INTEGER, PARAMETER :: n = 5, ne  = 7
   INTEGER :: ORDER( n )
   REAL ( KIND = wp ) :: B( n ), X( n )
!  REAL ( KIND = wp ) :: B2( n, 2 ), X2( n, 2 )
   INTEGER :: ROWS( n ), COLS( n )
   INTEGER, DIMENSION( ne ) :: row = (/ 1, 2, 2, 3, 3, 4, 5 /)
   INTEGER, DIMENSION( ne ) :: col = (/ 1, 1, 5, 2, 3, 3, 4 /)
   INTEGER, DIMENSION( n + 1 ) :: ptr = (/ 1, 2, 4, 6, 7, 8 /)
   REAL ( KIND = wp ), DIMENSION( ne ) ::                                      &
     val = (/ 2.0_wp, 3.0_wp, 6.0_wp, 4.0_wp,  1.0_wp, 5.0_wp, 1.0_wp /)
   REAL ( KIND = wp ), DIMENSION( n * n ) ::                                   &
     dense = (/ 2.0_wp, 0.0_wp, 0.0_wp, 0.0_wp, 0.0_wp,                        &
                3.0_wp, 0.0_wp, 0.0_wp, 0.0_wp, 6.0_wp,                        &
                0.0_wp, 4.0_wp, 1.0_wp, 0.0_wp, 0.0_wp,                        &
                0.0_wp, 0.0_wp, 5.0_wp, 0.0_wp, 0.0_wp,                        &
                0.0_wp, 0.0_wp, 0.0_wp, 1.0_wp, 0.0_wp /)
   REAL ( KIND = wp ), DIMENSION( n ) ::                                       &
     rhs = (/ 2.0_wp, 33.0_wp, 11.0_wp, 15.0_wp, 4.0_wp /)
   REAL ( KIND = wp ), DIMENSION( n ) ::                                       &
     SOL = (/ 1.0_wp, 2.0_wp, 3.0_wp, 4.0_wp, 5.0_wp /)

! Choose optional ordering
   DO i = 1, n
     ORDER( i ) = n - i + 1
   END DO

! Read matrix order and number of entries
!  DO type = 1, 1
!  DO type = 2, 2
   DO type = 1, 3
! Allocate arrays of appropriate sizes
     SELECT CASE( type )
     CASE ( 1 )
       write(6,"( ' coordinate storage ' )" )
       CALL SMT_put( matrix%type, 'COORDINATE', s )
       ALLOCATE( matrix%val( ne ), matrix%row( ne ), matrix%col( ne ) )
     CASE ( 2 )
       write(6,"( ' sparse by rows storage ' )" )
       ALLOCATE( matrix%val( ne ), matrix%ptr( n + 1 ), matrix%col( ne ) )
       CALL SMT_put( matrix%type, 'SPARSE_BY_ROWS', s )
     CASE ( 3 )
       write(6,"( ' dense storage ' )" )
       ALLOCATE( matrix%val( n * n ) )
       CALL SMT_put( matrix%type, 'DENSE', s )
     END SELECT
     matrix%m = n ; matrix%n = n
     DO ordering = 1, 3
!    DO ordering = 1, 1
       IF ( ordering == 1 ) THEN
         write(6,"( '  default ordering' )" )
       ELSE IF ( ordering == 2 ) THEN
         write(6,"( '  computed ordering' )" )
       ELSE
         write(6,"( '  provided ordering' )" )
       END IF
       WRITE( 6,"( '   solver  1 RHS  1 refine')")
!      DO solver = 1, 1
       DO solver = 1, 2
!      DO solver = 1, 5
! assign the matrix and right-hand side
         SELECT CASE( type )
         CASE ( 1 )
           matrix%ne = ne
           matrix%row = row
           matrix%col = col
           matrix%val = val
         CASE ( 2 )
           matrix%col = col
           matrix%val = val
           matrix%ptr = ptr
         CASE ( 3 )
           matrix%val = dense
         END SELECT
         B = rhs
! Initialize the structures
         IF ( solver == 1 ) THEN
           WRITE( 6, "( '     gls  ' )", advance = 'no' )
           CALL ULS_initialize( 'gls ', data, control, inform )
         ELSE IF ( solver == 2 ) THEN
           WRITE( 6, "( '     ma48 ' )", advance = 'no' )
           CALL ULS_initialize( 'ma48', data, control, inform )
         END IF
! Factorize
         CALL ULS_factorize( matrix, data, control, inform )
         IF ( inform%status == GALAHAD_unavailable_option ) THEN
           WRITE( 6, "( '  none ' )", advance = 'no' )
           WRITE( 6, "( '' )" )
           CYCLE
         ELSE IF ( inform%status < 0 ) THEN
           WRITE( 6, "( '  fail in factorize ' )", advance = 'no' )
           WRITE( 6, "( '' )" )
           CYCLE
         END IF
! Solve without refinement
         CALL ULS_solve( matrix, B, X, data, control, inform, .FALSE. )
         IF ( MAXVAL( ABS( X( 1 : n ) - SOL( 1: n ) ) )                        &
                 <= EPSILON( 1.0_wp ) ** 0.5 ) THEN
           WRITE( 6, "( '   ok  ' )", advance = 'no' )
         ELSE
           WRITE( 6, "( '  fail ' )", advance = 'no' )
         END IF
! Perform one refinement
         control%max_iterative_refinements = 1
         CALL ULS_solve( matrix, B, X, data, control, inform, .FALSE. )
         IF ( MAXVAL( ABS( X( 1 : n ) - SOL( 1: n ) ) )                        &
                 <= EPSILON( 1.0_wp ) ** 0.5 ) THEN
           WRITE( 6, "( '    ok  ' )", advance = 'no' )
         ELSE
           WRITE( 6, "( '  fail  ' )", advance = 'no' )
         END IF
! Solve multiple RHS without refinement
!         B2( : , 1 ) = B ; B2( : , 2 ) = B
!         X2 = B2
!         control%max_iterative_refinements = 0
!         CALL ULS_solve( matrix, B2, X2, data, control, inform, .FALSE. )
!         IF ( MAXVAL( ABS( X2( 1 : n, 1 ) - SOL( 1 : n ) ) )                  &
!                 <= EPSILON( 1.0_wp ) ** 0.5 .AND.                            &
!              MAXVAL( ABS( X2( 1 : n, 2 ) - SOL( 1 : n ) ) )                  &
!                 <= EPSILON( 1.0_wp ) ** 0.5 ) THEN
!           WRITE( 6, "( '      ok  ' )", advance = 'no' )
!         ELSE
!           WRITE( 6, "( '     fail ' )", advance = 'no' )
!         END IF
! Perform one refinement
!         control%max_iterative_refinements = 1
!         CALL ULS_solve( matrix, B2, X2, data, control, inform, .FALSE. )
!         IF ( MAXVAL( ABS( X2( 1 : n, 1 ) - SOL( 1 : n ) ) )                  &
!                 <= EPSILON( 1.0_wp ) ** 0.5 .AND.                            &
!              MAXVAL( ABS( X2( 1 : n, 2 ) - SOL( 1 : n ) ) )                  &
!                 <= EPSILON( 1.0_wp ) ** 0.5 ) THEN
!           WRITE( 6, "( '       ok  ' )", advance = 'no' )
!         ELSE
!           WRITE( 6, "( '      fail ' )", advance = 'no' )
!         END IF
         CALL ULS_enquire( data, inform, ROWS, COLS )
         CALL ULS_terminate( data, control, inform )
         WRITE( 6, "( '' )" )
       END DO
     END DO
     SELECT CASE( type )
     CASE ( 1 )
       DEALLOCATE( matrix%val, matrix%row, matrix%col )
     CASE ( 2 )
       DEALLOCATE( matrix%val, matrix%ptr, matrix%col )
     CASE ( 3 )
       DEALLOCATE( matrix%val )
     CASE ( 4 )
       DEALLOCATE( matrix%val )
     END SELECT
   END DO
!stop
! Test error returns
   WRITE( 6, "( ' error tests' )" )
   WRITE( 6, "( '   solver     -3   -26')" )
   DO solver = 1, 2
! Initialize the structures
     IF ( solver == 1 ) THEN
       WRITE( 6, "( '     gls  ' )", advance = 'no' )
       CALL ULS_initialize( 'gls ', data, control, inform )
     ELSE IF ( solver == 2 ) THEN
       WRITE( 6, "( '     ma48 ' )", advance = 'no')
       CALL ULS_initialize( 'ma48', data, control, inform )
     END IF
     control%error = - 1 ; control%warning = - 1
     control%out = - 1 ; control%print_level = - 1
     matrix%n = 0 ;  matrix%m = 0 ; matrix%ne = 0
     CALL SMT_put( matrix%type, 'COORDINATE', s )
     ALLOCATE( matrix%val( 0 ), matrix%row( 0 ), matrix%col( 0 ) )
! Factorize
     CALL ULS_factorize( matrix, data, control, inform )
     WRITE( 6, "( I6 )" ) inform%status
     DEALLOCATE( matrix%val, matrix%row, matrix%col )
     CALL ULS_terminate( data, control, inform )
   END DO
   WRITE( 6, "( '  unknown ' )", advance = 'no' )
   CALL ULS_initialize( 'unknown_solver', data, control, inform )
   WRITE( 6, "( 6X, I6 )" ) inform%status
   CALL ULS_terminate( data, control, inform )
   STOP
   END PROGRAM GALAHAD_ULS_TEST_PROGRAM

File: src/uls/C/ulst.c
/* ulst.c */
/* Full test for the ULS C interface using C sparse matrix indexing */

#include <stdio.h>
#include <math.h>
#include <stdlib.h>
#include <float.h>
#include "galahad_precision.h"
#include "galahad_cfunctions.h"
#include "galahad_uls.h"

ipc_ maxabsarray(rpc_ a[], ipc_ n, rpc_ *maxabs);

int main(void) {

    // Derived types
    void *data;
    struct uls_control_type control;
    struct uls_inform_type inform;

    // Set problem data
    ipc_ m = 5; // column dimension of A
    ipc_ n = 5; // column dimension of A
    ipc_ ne = 7; // number of entries of A
    ipc_ dense_ne = 25; // number of elements of A as a dense matrix
    ipc_ row[] = {0, 1, 1, 2, 2, 3, 4}; // row indices
    ipc_ col[] = {0, 0, 4, 1, 2, 2, 3}; // column indices
    ipc_ ptr[] = {0, 1, 3, 5, 6, 7}; // pointers to indices
    rpc_ val[] = {2.0, 3.0, 6.0, 4.0, 1.0, 5.0, 1.0}; // values
    rpc_ dense[] = {2.0, 0.0, 0.0, 0.0, 0.0, 3.0, 0.0, 0.0, 0.0, 6.0,
                      0.0, 4.0, 1.0, 0.0, 0.0, 0.0, 0.0, 5.0, 0.0, 0.0,
                      0.0, 0.0, 0.0, 1.0, 0.0};
    rpc_ rhs[] = {2.0, 33.0, 11.0, 15.0, 4.0};
    rpc_ rhst[] = {8.0, 12.0, 23.0, 5.0, 12.0};
    rpc_ sol[] = {1.0, 2.0, 3.0, 4.0, 5.0};
    ipc_ i, status;
    rpc_ x[n];
    rpc_ error[n];
    _Bool trans;

    rpc_ norm_residual;
    rpc_ good_x = pow( DBL_EPSILON, 0.3333 );

    printf(" C sparse matrix indexing\n\n");

    printf(" basic tests of storage formats\n\n");

    printf(" storage          RHS   refine   RHST  refine\n");

    for( ipc_ d=1; d <= 3; d++){
        // Initialize ULS - use the gls solver
        uls_initialize( "getr", &data, &control, &status );

        // Set user-defined control options
        control.f_indexing = false; // Fortran sparse matrix indexing

        switch(d){ // import matrix data and factorize
            case 1: // sparse co-ordinate storage
                printf(" coordinate     ");
                uls_factorize_matrix( &control, &data, &status, m, n,
                                    "coordinate", ne, val, row, col, NULL );
                break;
            case 2: // sparse by rows
                printf(" sparse by rows ");
                uls_factorize_matrix( &control, &data, &status, m, n,
                                    "sparse_by_rows", ne, val, NULL, col, ptr );
                break;
            case 3: // dense
                printf(" dense          ");
                uls_factorize_matrix( &control, &data, &status, m, n, "dense",
                                      dense_ne, dense, NULL, NULL, NULL );
                break;
            }

        // Set right-hand side and solve the system A x = b
        for(i=0; i<n; i++) x[i] = rhs[i];
        trans = false;
        uls_solve_system( &data, &status, m, n, x, trans );
        uls_information( &data, &inform, &status );

        if(inform.status == 0){
          for(i=0; i<n; i++) error[i] = x[i]-sol[i];
          status = maxabsarray( error, n, &norm_residual );
          if(norm_residual < good_x){
            printf("   ok  ");
          }else{
            printf("  fail ");
          }
        }else{
            printf(" ULS_solve exit status = %1" i_ipc_ "\n", inform.status);
        }
        // printf("sol: ");
        // for( ipc_ i = 0; i < n; i++) printf("%f ", x[i]);

        // resolve, this time using iterative refinement
        control.max_iterative_refinements = 1;
        uls_reset_control( &control, &data, &status );
        for(i=0; i<n; i++) x[i] = rhs[i];
        uls_solve_system( &data, &status, m, n, x, trans );
        uls_information( &data, &inform, &status );

        if(inform.status == 0){
          for(i=0; i<n; i++) error[i] = x[i]-sol[i];
          status = maxabsarray( error, n, &norm_residual );
          if(norm_residual < good_x){
            printf("    ok  ");
          }else{
            printf("   fail ");
          }
        }else{
            printf(" ULS_solve exit status = %1" i_ipc_ "\n", inform.status);
        }

        // Set right-hand side and solve the system A^T x = b
        for(i=0; i<n; i++) x[i] = rhst[i];
        trans = true;
        uls_solve_system( &data, &status, m, n, x, trans );
        uls_information( &data, &inform, &status );

        if(inform.status == 0){
          for(i=0; i<n; i++) error[i] = x[i]-sol[i];
          status = maxabsarray( error, n, &norm_residual );
          if(norm_residual < good_x){
            printf("   ok  ");
          }else{
            printf("  fail ");
          }
        }else{
            printf(" ULS_solve exit status = %1" i_ipc_ "\n", inform.status);
        }
        // printf("sol: ");
        // for( ipc_ i = 0; i < n; i++) printf("%f ", x[i]);

        // resolve, this time using iterative refinement
        control.max_iterative_refinements = 1;
        uls_reset_control( &control, &data, &status );
        for(i=0; i<n; i++) x[i] = rhst[i];
        uls_solve_system( &data, &status, m, n, x, trans );
        uls_information( &data, &inform, &status );

        if(inform.status == 0){
          for(i=0; i<n; i++) error[i] = x[i]-sol[i];
          status = maxabsarray( error, n, &norm_residual );
          if(norm_residual < good_x){
            printf("    ok  ");
          }else{
            printf("   fail ");
          }
        }else{
            printf(" ULS_solve exit status = %1" i_ipc_ "\n", inform.status);
        }

        // Delete internal workspace
        uls_terminate( &data, &control, &inform );
        printf("\n");
    }
}

ipc_ maxabsarray(rpc_ a[], ipc_ n, rpc_ *maxabs)
 {
    ipc_ i;
    rpc_ b, max;
    max=abs(a[0]);
    for(i=1; i<n; i++)
    {
        b = fabs(a[i]);
	if(max<b)
          max=b;
    }
    *maxabs=max;
    return 0;
 }
